### PR TITLE
refactor dataTrieTracker

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- There were some duplicated mocks
- TrackableDataTrie was exported
  
## Proposed Changes
- Remove duplicated mocks
- Unexport trackableDataTrie and make in inaccessible form outside the account

## Testing procedure
- Normal test

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -7,8 +7,8 @@ require (
 	github.com/ElrondNetwork/arwen-wasm-vm/v1_3 v1.3.42-0.20220729115131-85ecca868e90
 	github.com/ElrondNetwork/arwen-wasm-vm/v1_4 v1.4.59-0.20220729115431-a6c93119bdda
 	github.com/ElrondNetwork/covalent-indexer-go v1.0.6
-	github.com/ElrondNetwork/elastic-indexer-go v1.2.41
-	github.com/ElrondNetwork/elrond-go-core v1.1.19
+	github.com/ElrondNetwork/elastic-indexer-go v1.2.42-0.20220921110140-860b4b8c7fc3
```
Comment (iulianpascalau) [inline go.mod:10]: do not forget about proper releases

File: common/interface.go [hunk 2]
```diff
@@ -50,6 +50,13 @@ type KeyBuilder interface {
 	Clone() KeyBuilder
 }
 
+// DataTrie is an interface that declares the methods used for dataTries
+type DataTrie interface {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/interface.go:54]: Done.

File: state/journalEntries.go [hunk 3]
```diff
@@ -187,16 +188,21 @@ func NewJournalEntryDataTrieUpdates(trieUpdates map[string][]byte, account baseA
 
 // Revert applies undo operation
 func (jedtu *journalEntryDataTrieUpdates) Revert() (vmcommon.AccountHandler, error) {
+	trie, ok := jedtu.account.DataTrie().(common.Trie)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/journalEntries.go:191]: There are only 2 casts in productive code (the rest are in tests). The idea behind this was to not give access directly to the trie (although this is possible through casting).

File: state/trackableDataTrie.go [hunk 4]
```diff
@@ -85,16 +76,51 @@ func (tdaw *TrackableDataTrie) SaveKeyValue(key []byte, value []byte) error {
...
+// SaveDirtyData saved the dirty data to the trie
+func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) (map[string][]byte, error) {
+	if len(tdaw.dirtyData) == 0 {
+		return map[string][]byte{}, nil
+	}
+
+	if check.IfNil(tdaw.tr) {
+		newDataTrie, err := mainTrie.Recreate(make([]byte, 0))
+		if err != nil {
+			return map[string][]byte{}, err
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:97]: Done.

File: state/trackableDataTrie.go [hunk 5]
```diff
@@ -85,16 +76,51 @@ func (tdaw *TrackableDataTrie) SaveKeyValue(key []byte, value []byte) error {
...
+		oldValues[k] = val
+
+		err = tdaw.tr.Update([]byte(k), v)
+		if err != nil {
+			return oldValues, err
+		}
+	}
+
+	tdaw.dirtyData = make(map[string][]byte)
+	return oldValues, nil
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:120]: Yes, this is used only for journalizing. (accountsDB -> saveDataTrie())

File: state/trackableDataTrie.go [hunk 6]
```diff
@@ -2,40 +2,31 @@ package state
...
+func NewTrackableDataTrie(identifier []byte, tr common.Trie) *trackableDataTrie {
+	return &trackableDataTrie{
 		tr:         tr,
 		dirtyData:  make(map[string][]byte),
 		identifier: identifier,
 	}
 }
 
-// ClearDataCaches empties the dirtyData map and original map
-func (tdaw *TrackableDataTrie) ClearDataCaches() {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:26]: Correct. It is called inside SaveDirtyData(), but we don't need a method for this.

File: go.mod (hunk 1 above)
Comment (BeniaminDrasovean) [reply go.mod:10]: Will do.

File: state/journalEntries.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/journalEntries.go:191]: ok, as agreed, we will keep this as it is

File: state/trackableDataTrie.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/trackableDataTrie.go:120]: ok, will remain as it is

File: state/accountsDB.go [hunk 7]
```diff
@@ -498,67 +498,34 @@ func (adb *AccountsDB) loadDataTrie(accountHandler baseAccountHandler) error {
...
-	trackableDataTrie.ClearDataCaches()
-
-	log.Trace("accountsDB.SaveDataTrie",
-		"address", hex.EncodeToString(accountHandler.AddressBytes()),
-		"new root hash", accountHandler.GetRootHash(),
-	)
 
 	if check.IfNil(adb.dataTries.Get(accountHandler.AddressBytes())) {
-		adb.dataTries.Put(accountHandler.AddressBytes(), accountHandler.DataTrie())
+		trie, ok := accountHandler.DataTrie().(common.Trie)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/accountsDB.go:522]: will remain as it is

File: node/node.go [hunk 8]
```diff
@@ -1172,7 +1172,7 @@ func (n *Node) getAccountRootHashAndVal(address []byte, accBytes []byte, key []b
 		return nil, nil, fmt.Errorf("empty dataTrie rootHash")
 	}
 
-	retrievedVal, err := userAccount.RetrieveValueFromDataTrieTracker(key)
+	retrievedVal, err := userAccount.RetrieveValue(key)
```
Comment (bogdan-rosianu) [inline node/node.go:1175]: when system-testing this branch, please make sure you test the `/address/erd1.../esdt` endpoint. it should validate that the affected functionalities (save key value, retrieve value) work the same

**Review by gabi-vuls — DISMISSED** [review]
System test passed.

**Review by iulianpascalau — COMMENTED** [review]
Just the proper release comment remains

File: node/node.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply node/node.go:1175]: It works as expected.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**